src/libstd/sys/unix/mutex.rs

```diff
@@ -67,3 +68,38 @@ impl Mutex {
         debug_assert!(r == 0 || r == libc::EINVAL);
     }
 }
+
+pub struct ReentrantMutex { inner: UnsafeCell<ffi::pthread_mutex_t> }
+
+unsafe impl Send for ReentrantMutex {}
+unsafe impl Sync for ReentrantMutex {}
+
+impl ReentrantMutex {
+    pub unsafe fn new() -> ReentrantMutex {
+        let mut mutex: ffi::pthread_mutex_t = mem::uninitialized();
+        let mut attr: ffi::pthread_mutexattr_t = mem::uninitialized();
+        debug_assert_eq!(ffi::pthread_mutexattr_init(&mut attr as *mut _), 0);
```

Comment:
I think you'll want to lift out these expressions from `debug_assert_eq!` as otherwise they'll be omitted in release builds.